core/service.go

```diff
@@ -295,21 +303,57 @@ func (s *Service) handleReceivedMessage(msg network.Message) (err error) {
 // announce messages (block announce messages include the header but the full
 // block is required to execute `core_execute_block`).
 func (s *Service) ProcessBlockAnnounceMessage(msg network.Message) error {
+	blockAnnounceMessage, ok := msg.(*network.BlockAnnounceMessage)
+	if !ok {
+		return errors.New("could not cast p2p.Message to BlockAnnounceMessage")
+	}
 
-	// TODO: check if we should send block request message
-
-	// TODO: update message properties and use generated id
-	blockRequest := &network.BlockRequestMessage{
-		ID:            1,
-		RequestedData: 2,
-		StartingBlock: []byte{},
-		EndBlockHash:  optional.NewHash(true, common.Hash{}),
-		Direction:     1,
-		Max:           optional.NewUint32(false, 0),
+	if s.blockState == nil {
+		return errors.New("ProcessBlockAnnounceMessage error: blockState is nil")
 	}
 
-	// send block request message to network service
-	s.msgSend <- blockRequest
+	latestBlockNum := s.blockState.LatestHeader().Number
+	messageBlockNum := blockAnnounceMessage.Number
+
+	// check if we should send block request message
+	if latestBlockNum.Cmp(messageBlockNum) == 1 {
+		// BlockAnnounceMessage
+		blockAnnounce := &network.BlockAnnounceMessage{
+			ParentHash:     blockAnnounceMessage.ParentHash,
+			Number:         messageBlockNum,
+			StateRoot:      blockAnnounceMessage.StateRoot,
+			ExtrinsicsRoot: blockAnnounceMessage.ExtrinsicsRoot,
+			Digest:         blockAnnounceMessage.Digest,
+		}
+
+		// send blockAnnounce message to network service
+		log.Debug("send blockAnnounce message to network service")
+		s.msgSend <- blockAnnounce
+
+	} else {
+
+		//generate random ID
+		s1 := rand.NewSource(uint64(time.Now().UnixNano()))
+		seed := rand.New(s1).Uint64()
+		randomID := mrand.New(mrand.NewSource(int64(seed))).Uint64()
+
+		blockRequest := &network.BlockRequestMessage{
+			ID:            randomID, // random
+			RequestedData: 2,
+			StartingBlock: []byte{},
```

Comment:
starting block should be `byte(0)` + block hash or `byte(1)` + block Number

Reply:
sure